pysmt/optimization/optimizer.py

```diff
@@ -168,15 +168,15 @@ def optimize(self, cost_function, strategy='ub',
             else:
                 # Check if there is a model </<= bound
                 check_result = self._optimization_check_progress(client_data,
-                                                                 cost_function,
+                                                                 goal.minimize_term(),
```

Comment:
You need to check that goal is a Minimization or Maximization goal, otherwise throw a meaningful exception